src/api_sketch.jl

```diff
@@ -0,0 +1,211 @@
+    # probably needs some other stuff...
+end
+# same idea here...copy from weave stuff...
+# a note: I think it's more elegant (and probably efficient?) if pair features assert
+# always arrays where the first two indices are atom indices and any further are indexing
+# into the feature if it is itself higher-dimensional. This is not how the weave featurizer
+# currently works...
+
+#=
+This is just an example of another sort of feature one might add, but it's not needed/used
+currently.
+=#
+struct StructureFeat{Tn,Te}<:AbstractFeature{Tn,Te}
+    name
+    encode_f
+    decode_f
+    # etc...
+end
+
+#= FEATURIZATION OBJECTS
+All such objects should define at least one list of <:AbstractFeature objects and a `combine`
+function that specifies how to put them together for eventual attachment to an atoms 
+object. 
+=#
+
+abstract type AbstractFeaturization end
+
+struct GraphNodeFeaturization <: AbstractFeaturization
+    atom_feats::Vector{AtomFeat}
+    feature_vectors::Dict{String,Vector{Float32}} # map from element symbol to vector
+    combine
+end
+
+struct WeaveFeaturization <: AbstractFeaturization
+    atom_feats::Vector{AtomFeat}
+    pair_feats::Vector{PairFeat}
+    combine
+end
+
+
+#= ATOMS OBJECTS
+
+AtomGraph shouldn't have to change too much from its current incarnation. WeaveMol is 
+not currently a thing that can be directly fed into a model, so that will have to get
+updated. Both will need some way to specify which types of features can be attached 
+to them, maybe this is a place for the holy trait design pattern?
+
+Example: AtomGraph can take ElementFeat and ComputedAtomFeat but not PairFeat or StructureFeat, WeaveMol can take ElementFeat, ComputedAtomFeat, and PairFeat but also not StructureFeat
+=#
+
+abstract type AbstractAtoms end
+
+# it is kind of cool to subtype this from LightGraphs so all that machinery "just works"
+# that being said, all of it would "just work" by just pulling out ag.graph so may be
+# more cool than actually practical, hence I've done the abstract type here
+mutable struct AtomGraph <: AbstractAtoms
+    graph::SimpleWeightedGraph{Int32,Float32}
+    elements::Vector{String}
+    lapl::LightGraphs.LinAlg.NormalizedLaplacian
+    features::Matrix{Float32} # if we add edge features this type will have to relax
+    featurization::GraphNodeFeaturization
+    id::String # or maybe we let it be a number too?
+end
+
+mutable struct WeaveMol <: AbstractAtoms
+    smiles::String
+    elements::Vector{String}
+    features::Tuple{SomethingOrOther} # I need to look more carefully to figure this out heh
+    featurization::WeaveFeaturization
+    id # probably makes sense to have this in addition to smiles? Maybe?
+end
+
+# constructors etc. as before
+# maybe some cutesy stuff like dispatching things like length as length(elements)
+# pretty printing stuff
+
+# generic featurize...there's probably a better way to write this...
+function featurize!(a<:AbstractAtoms, f<:AbstractFeaturization)
```

Comment:
Could we revisit this or add a TODO? I wouldn't want to hold progress for this, but yeah I agree we should make it neater.